### PR TITLE
First pass of FromPerlKV derive

### PR DESCRIPTION
**Pull request by dnorman**
First swing at FromPerlKV derive. Tests passing: https://travis-ci.org/dnorman/perl-xs/builds/301790192

I'm pretty inexperienced with proc_macros, so I'm sure there's lots of ways this can be improved. Additionally, I couldn't figure out how to each over HV contents, so hashrefs are currently unhandled.

Option<T> on the struct is handled differently versus T. The end goal here is to loosely reproduce the behavior of a moose constructor. Feedback and improvements greatly welcomed.

```
#[derive(FromPerlKV)]
struct Test {
    foo: Option<String>,
    bar: String,
    #[perlxs(key = "-meow")]
    meow: bool
};

xs! {
    package Test;

    sub new(ctx, class: String) {
        println!("new {:?}", class);
        let test = Test::from_perl_kv(&mut ctx, 1); # offset to avoid $_[0]
    }
}

...
Test->new( bar => "Something", -meow => 1 ) #does not error for missing foo
```

### CODE REVIEW

**Comment by vickenty** [comment]
First of all, thanks! This is awesome.

> Additionally, I couldn't figure out how to each over HV contents, so hashrefs are currently unhandled. 

Adding methods to iterate hashes is easy, but I think it might be better to leave hashrefs out of this trait, and make it possible to derive `TryFromSV` instead. This would allow users to pick behaviors, and a simple wrapper wrapper would be able to pick `TryFromSV` or `FromPerlKV` depending on number of arguments.

> Option on the struct is handled differently versus T. The end goal here is to loosely reproduce the  behavior of a moose constructor.

I think this would be fine if we tighten up `TryFromSV` implementations to only accept valid scalars for desired type; then invalid values would fail the conversions and produce a proper error instead of a nonsense value with a warning.

Other than that, I think this code is in good shape and can be merged after a cleanup. :)

File: perlxs_derive/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+            i += 2;
+        };
+
+        Self{
+            #(#paramvars,)*
+        }
+    };
+    
+    let impl_block = quote! {
+        impl #impl_generics #ident #ty_generics #where_clause {
```
Comment (vickenty) [inline perlxs_derive/src/lib.rs]: This creates an inherent impl instead of implementing `FromPerlKV` trait. :)

File: perlxs_derive/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+            paramvars.push(
+                quote! {
+                    #ident: #var
+                }
+            );
+        }else{
+            let err_omitted = field.err_omitted();
+
+            // I'm not super in love with using expect for this.
+            // There's got to be a better way
```
Comment (vickenty) [inline perlxs_derive/src/lib.rs]: `from_perl_kv` could return a `Result`, like `TryFromSV` does.

**Comment by dnorman** [comment]
Thanks for the feedback. Will take a crack at it this weekend and update the PR :)

**Comment by dnorman** [comment]
Hey @vickenty, how does this look now?

File: src/error.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+/// Partial error instantiating a rust struct from a perl stack
+#[derive(Debug)]
+pub enum ToStructErrPart {
+    /// A list of keys one of which must be specified
+    OmittedKey(&'static [&'static str]),
+    /// A key for which a value was not specified
+    OmittedValue(&'static str),
+    ///
+    KeyParseFail{
```
Comment (vickenty) [inline src/error.rs]: Missing doc-string.

File: perlxs_derive/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+                    match &*key {
+                        #(#matchparts,)*
+                        &_ => {
+                            // Unknown key. Should we warn?
+                        }
+                    }
+                },
+                Err(e) => {
+                    //TODO
+                    errors.push(_perlxs::error::ToStructErrPart::KeyParseFail{ offset: i, ty: "String", error: e.to_string() });
```
Comment (vickenty) [inline perlxs_derive/src/lib.rs]: Something needs to be done here?

**Comment by vickenty** [comment]
Sorry for taking so long to review this.

This looks very good, I think it's almost ready to be merged, pending only a cleanup - there is some commented out code that I think can be removed, and a few other nitpicks that I've left inline comments for. Once merged, I plan on doing a global rustfmt run, so no need to worry about formatting (I think old code would benefit greatly from it too).

Thanks again for doing this amazing work, this is very cool indeed. :)

File: perlxs_derive/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    let from_kv_stack = quote!{
+
+        let mut i = offset;
+        while let Some(sv_res) = ctx.st_try_fetch::<String>(i) {
+            match sv_res {
+                Ok(key) => {
+                    match &*key {
+                        #(#matchparts,)*
+                        &_ => {
+                            // Unknown key. Should we warn?
```
Comment (vickenty) [inline perlxs_derive/src/lib.rs]: A warning here would make sense I think, but perl-xs lacks bindings for that yet. I am okay with leaving it as is for now and revisiting later, when `warn` is properly implemented.

**Comment by dnorman** [comment]
Should be a bit tidier now @vickenty 
Try try again? :)

**Comment by vickenty** [comment]
Thanks!

**Pull request merged**